training/src/main/scala/com/airbnb/aerosolve/training/FullRankLinearTrainer.scala

```diff
@@ -0,0 +1,220 @@
+    })
+    log.info("Sum of Gradient L2 norms = " + gradientNorm)
+    log.info("Num active features = " + featureCount)
+  }
+  
+  def softmaxGradient(sc : SparkContext,
+                      options : FullRankLinearTrainerOptions,
+                      model : FullRankLinearModel,
+                      pointwise : RDD[Example]) : Array[((String, String), GradientContainer)] = {
+    val modelBC = sc.broadcast(model)
+    
+    pointwise
+    .mapPartitions(partition => {
+      val model = modelBC.value
+      val labelToIdx = model.getLabelToIndex()
+      val dim = model.getLabelDictionary.size()
+      val gradient = scala.collection.mutable.HashMap[(String, String), GradientContainer]()
+      val weightVector = model.getWeightVector()
+
+      partition.foreach(examples => {
+        val flatFeatures = Util.flattenFeature(examples.example.get(0))
+        val labels = flatFeatures.get(options.rankKey)
+        if (labels != null) {
+          val posLabels = labels.keySet().asScala
+          val scores = model.scoreFlatFeature(flatFeatures)
+          // Convert to multinomial using softmax.
+          scores.softmax()
+          // The importance is prob - 1 for positive labels, prob otherwise.
+          for (posLabel <- posLabels) {
+            val posIdx = labelToIdx.get(posLabel)
+            if (posIdx != null) {
+              scores.values(posIdx) -= 1
+            }
+          }
+          // Gradient is importance * feature value
+          for (family <- flatFeatures) {
+            for (feature <- family._2) {
+              val key = (family._1, feature._1)
+              // We only care about features in the model.
+              if (weightVector.containsKey(key._1) && weightVector.get(key._1).containsKey(key._2)) {
+                val featureVal = feature._2
+                val gradContainer = gradient.getOrElse(key,
+                                                       GradientContainer(new FloatVector(dim), 0.0))
+                gradContainer.grad.multiplyAdd(featureVal.toFloat, scores)
+                gradient.put(key,
+                             GradientContainer(gradContainer.grad,
+                             gradContainer.featureSquaredSum + featureVal * featureVal
+                ))
+              }
+            }
+          }
+        }
+      })
+      gradient.iterator
+    })
+    .reduceByKey((a, b) => sumGradients(a,b))
+    .collect
+  }
+
+  def parseTrainingOptions(config : Config) : FullRankLinearTrainerOptions = {
+ 
+    FullRankLinearTrainerOptions(
+        loss = config.getString("loss"),
+        iterations = config.getInt("iterations"),
+        rankKey = config.getString("rank_key"),
+        lambda = config.getDouble("lambda"),
+        minCount = config.getInt("min_count")
+    )    
+  }
+  
+  def setupModel(options : FullRankLinearTrainerOptions, pointwise : RDD[Example]) : FullRankLinearModel = {
+    val stats = TrainingUtils.getFeatureStatistics(options.minCount, pointwise)
```

Comment:
basically, we only need the count of samples from each class, i.e. count for the feature that represents label, but this method will compute min, max, variance for all features which seems a lot of waste of computation